File: servers/compass/src/tools/project-tools.ts
import { z } from "zod";
import { join } from "https://deno.land/std@0.208.0/path/mod.ts";
import { exists } from "https://deno.land/std@0.208.0/fs/mod.ts";
// Temporarily implement writeFile locally until import issues are resolved
async function writeFile(args: { path: string; content: string }) {
  const dir = args.path.substring(0, args.path.lastIndexOf("/"));
  await Deno.mkdir(dir, { recursive: true });
  await Deno.writeTextFile(args.path, args.content);
}
import type { MCPTool } from "@packages/mcp-core/mod.ts";

// Schema definitions
const InitProjectInputSchema = z.object({
  name: z.string().describe("Name of the project to initialize"),
  projectsDir: z.string().optional().describe("Projects directory (for testing)"),
});

const InitProjectOutputSchema = z.object({
  success: z.boolean(),
  project_name: z.string(),
  project_path: z.string(),
  message: z.string(),
  next_steps: z.array(z.string()),
});

const ListProjectsInputSchema = z.object({
  projectsDir: z.string().optional().describe("Projects directory (for testing)"),
});

const ListProjectsOutputSchema = z.object({
  projects: z.array(z.object({
    name: z.string(),
    path: z.string(),
    status: z.string(),
    current_phase: z.string(),
    created: z.string(),
  })),
  total: z.number(),
});

const GetProjectStatusInputSchema = z.object({
  project_name: z.string().describe("Name of the project"),
  projectsDir: z.string().optional().describe("Projects directory (for testing)"),
});

const GetProjectStatusOutputSchema = z.object({
  project_name: z.string(),
  current_phase: z.string(),
  status_summary: z.string(),
  exploration_tasks: z.array(z.object({
    text: z.string(),
    completed: z.boolean(),
  })),
  next_action: z.string(),
  recommendations: z.array(z.string()),
});

// Tool definitions
export const projectTools: MCPTool[] = [
  {
    name: "init_project",
    description: "Initialize a new Compass project with systematic methodology structure",
    inputSchema: InitProjectInputSchema,
    outputSchema: InitProjectOutputSchema,
  },
  {
    name: "list_projects",
    description: "List all available Compass projects with status information",
    inputSchema: ListProjectsInputSchema,
    outputSchema: ListProjectsOutputSchema,
  },
  {
    name: "get_project_status",
    description: "Get detailed status of a specific project including tasks and recommendations",
    inputSchema: GetProjectStatusInputSchema,
    outputSchema: GetProjectStatusOutputSchema,
  },
];

// Helper functions
async function getProjectsDirectory(customDir?: string): Promise<string> {
  if (customDir) return customDir;
  
  const homeDir = Deno.env.get("HOME") || Deno.env.get("USERPROFILE");
  if (!homeDir) {
    throw new Error("Unable to determine user home directory");
  }
  
  const projectsDir = join(homeDir, "CompassProjects");
  await Deno.mkdir(projectsDir, { recursive: true });
  return projectsDir;
}

// Tool handlers
export const projectHandlers = {
  init_project: async (args: z.infer<typeof InitProjectInputSchema>) => {
    const { name, projectsDir: customDir } = args;
    
    try {
      const projectsDir = await getProjectsDirectory(customDir);
      const projectPath = join(projectsDir, name);
      
      // Check if project already exists
      if (await exists(projectPath)) {
        return {
          success: false,
          project_name: name,
          project_path: projectPath,
          message: `Project '${name}' already exists`,
          next_steps: [],
        };
      }
      
      // Create project structure
      await Deno.mkdir(projectPath, { recursive: true });
      await Deno.mkdir(join(projectPath, "exploration"), { recursive: true });
      await Deno.mkdir(join(projectPath, "specification"), { recursive: true });
      await Deno.mkdir(join(projectPath, "execution"), { recursive: true });
      await Deno.mkdir(join(projectPath, "feedback"), { recursive: true });
      
      // Create metadata
      const metadata = {
        name,
        created: new Date().toISOString(),
        currentPhase: "exploration",
        status: "active",
        sessionCount: 0,
      };
      
      await writeFile({
        path: join(projectPath, ".compass.json"),
        content: JSON.stringify(metadata, null, 2),
      });
      
      // Create initial tasks file
      const tasksContent = `# ${name} - Project Tasks

## Phase 1: Exploration
- [ ] Complete initial problem exploration

## Phase 2: Specification
*Tasks will be generated after exploration completion*

## Phase 3: Execution
*Tasks will be generated after specification completion*

## Phase 4: Feedback & Learning
*Tasks will be generated after execution completion*
`;
      
      await writeFile({
        path: join(projectPath, "tasks.md"),
        content: tasksContent,
      });
      
      return {
        success: true,
        project_name: name,
        project_path: projectPath,
        message: `COMPASS PROJECT INITIALIZED

Project '${name}' created with systematic 4-phase methodology:
• Phase 1: EXPLORATION (Current) - Understand the problem space  
• Phase 2: SPECIFICATION - Transform insights into structured requirements
• Phase 3: EXECUTION - Implement solutions based on solid understanding
• Phase 4: FEEDBACK & LEARNING - Capture knowledge for future projects

METHODOLOGY ENFORCEMENT: You must complete exploration before moving to specification or implementation.`,
        next_steps: [
          "Project structure created with 4-phase methodology",
          "Initial exploration tasks generated",
          "NEXT ACTION: Begin systematic exploration phase",
          `Project location: ${projectPath}`,
        ],
      };
    } catch (error) {
      return {
        success: false,
        project_name: name,
        project_path: "",
        message: `Failed to initialize project: ${error instanceof Error ? error.message : String(error)}`,
        next_steps: [],
      };
    }
  },
  
  list_projects: async (args: z.infer<typeof ListProjectsInputSchema>) => {
    const { projectsDir: customDir } = args;
    
    try {
      const projectsDir = await getProjectsDirectory(customDir);
      const projects = [];
      
      if (await exists(projectsDir)) {
        for await (const entry of Deno.readDir(projectsDir)) {
          if (entry.isDirectory) {
            const projectPath = join(projectsDir, entry.name);
            const metadataPath = join(projectPath, ".compass.json");
            
            if (await exists(metadataPath)) {
              const metadata = JSON.parse(await Deno.readTextFile(metadataPath));
              projects.push({
                name: entry.name,
                path: projectPath,
                status: metadata.status || "active",
                current_phase: metadata.currentPhase || "exploration",
                created: metadata.created || "unknown",
              });
            }
          }
        }
      }
      
      return {
        projects,
        total: projects.length,
      };
    } catch (error) {
      return {
        projects: [],
        total: 0,
      };
    }
  },
  
  get_project_status: async (args: z.infer<typeof GetProjectStatusInputSchema>) => {
    const { project_name, projectsDir: customDir } = args;
    
    try {
      const projectsDir = await getProjectsDirectory(customDir);
      const projectPath = join(projectsDir, project_name);
      
      if (!await exists(projectPath)) {
        throw new Error(`Project '${project_name}' not found`);
      }
      
      const metadata = JSON.parse(await Deno.readTextFile(join(projectPath, ".compass.json")));
      
      // TODO: Parse tasks from tasks.md
      const explorationTasks = [
        { text: "Complete initial problem exploration", completed: false },
      ];
      
      return {
        project_name,
        current_phase: metadata.currentPhase || "exploration",
        status_summary: "Exploration phase - Ready to begin",
        exploration_tasks: explorationTasks,
        next_action: "start_exploration",
        recommendations: ["Begin systematic exploration phase"],
      };
    } catch (error) {
      throw error;
    }
  },
};

File: servers/compass/src/project_tools.py
"""
Project management tools for Compass server
"""
import json
import logging
import os
from datetime import datetime
from pathlib import Path
from typing import List, Optional

from pydantic import BaseModel, Field

from core.types import MCPTool
from .project_manager import ProjectManager, ProjectMetadata


logger = logging.getLogger(__name__)

# Get workspace path from environment or use default
DEFAULT_WORKSPACE = os.getenv("COMPASS_WORKSPACE", "./workspace")


# Schema definitions
class InitProjectInputSchema(BaseModel):
    """Schema for initializing a project"""
    name: str = Field(description="Name of the project to initialize")
    projects_dir: Optional[str] = Field(default=DEFAULT_WORKSPACE, description="Projects directory")


class InitProjectOutputSchema(BaseModel):
    """Schema for init project output"""
    success: bool
    project_name: str
    project_path: str
    message: str
    next_steps: List[str]


class ProjectSummary(BaseModel):
    """Project summary for listing"""
    name: str
    path: str
    status: str
    current_phase: str
    created: str


class ListProjectsInputSchema(BaseModel):
    """Schema for listing projects"""
    projects_dir: Optional[str] = Field(default=DEFAULT_WORKSPACE, description="Projects directory")


class ListProjectsOutputSchema(BaseModel):
    """Schema for list projects output"""
    projects: List[ProjectSummary]
    total: int


class TaskSummary(BaseModel):
    """Task summary for status"""
    text: str
    completed: bool


class GetProjectStatusInputSchema(BaseModel):
    """Schema for getting project status"""
    project_name: str = Field(description="Name of the project")
    projects_dir: Optional[str] = Field(default=DEFAULT_WORKSPACE, description="Projects directory")


class GetProjectStatusOutputSchema(BaseModel):
    """Schema for project status output"""
    project_name: str
    current_phase: str
    status_summary: str
    exploration_tasks: List[TaskSummary]
    next_action: str
    recommendations: List[str]


# Tool definitions
project_tools: List[MCPTool] = [
    MCPTool(
        name="init_project",
        description="Initialize a new Compass project with systematic methodology structure",
        input_schema=InitProjectInputSchema.model_json_schema(),
        output_schema=InitProjectOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="list_projects",
        description="List all Compass projects with their current status",
        input_schema=ListProjectsInputSchema.model_json_schema(),
        output_schema=ListProjectsOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="get_project_status",
        description="Get detailed status and progress of a specific project",
        input_schema=GetProjectStatusInputSchema.model_json_schema(),
        output_schema=GetProjectStatusOutputSchema.model_json_schema(),
    ),
]


# Tool handlers
async def init_project_handler(args: dict) -> dict:
    """Initialize a new Compass project"""
    validated_args = InitProjectInputSchema(**args)
    
    try:
        projects_dir = Path(validated_args.projects_dir)
        projects_dir.mkdir(parents=True, exist_ok=True)
        
        project_path = projects_dir / validated_args.name
        
        if project_path.exists():
            return InitProjectOutputSchema(
                success=False,
                project_name=validated_args.name,
                project_path=str(project_path),
                message=f"Project '{validated_args.name}' already exists",
                next_steps=["Choose a different project name", "Use existing project"]
            ).model_dump()
        
        # Create project structure
        project_path.mkdir(parents=True, exist_ok=True)
        (project_path / "exploration").mkdir(exist_ok=True)
        (project_path / "specification").mkdir(exist_ok=True)
        (project_path / "execution").mkdir(exist_ok=True)
        (project_path / "feedback").mkdir(exist_ok=True)
        
        # Create project manager and initialize metadata
        project_manager = ProjectManager(str(project_path))
        metadata = ProjectMetadata(
            name=validated_args.name,
            created=datetime.now().isoformat(),
            current_phase="exploration",
            status="active",
            session_count=0
        )
        await project_manager.save_project_metadata(metadata)
        
        # Create initial tasks.md
        await project_manager.update_tasks_file(
            exploration_tasks=[],
            next_action="Start exploration with 'start_exploration' tool"
        )
        
        # Create questions.md for exploration guidance
        questions_content = f"""# {validated_args.name} - Exploration Questions

## Core Understanding Questions

- [ ] What is the fundamental problem or opportunity this project addresses?
- [ ] Who are the primary stakeholders and what are their needs?
- [ ] What are the success criteria and how will we measure them?
- [ ] What constraints and limitations exist?
- [ ] What assumptions are we making that need validation?

## Context & Environment

- [ ] What is the broader context this project operates within?
- [ ] What existing solutions or approaches have been tried?
- [ ] What resources and capabilities are available?
- [ ] What dependencies and external factors must be considered?

## Scope & Boundaries

- [ ] What is explicitly included in this project scope?
- [ ] What is explicitly excluded from this project scope?
- [ ] What are the minimum viable requirements?
- [ ] What are the nice-to-have enhancements?

## Risks & Challenges

- [ ] What are the primary risks and how can they be mitigated?
- [ ] What technical challenges need to be solved?
- [ ] What organizational or process challenges exist?
- [ ] What unknown unknowns might we discover?

---

*Use these questions to guide exploration conversations. Each question should be thoroughly explored before moving to specification phase.*
"""
        
        questions_path = project_path / "exploration" / "questions.md"
        questions_path.write_text(questions_content, encoding='utf-8')
        
        logger.info(f"Initialized new Compass project: {validated_args.name}")
        
        return InitProjectOutputSchema(
            success=True,
            project_name=validated_args.name,
            project_path=str(project_path),
            message=f"Successfully initialized Compass project '{validated_args.name}'",
            next_steps=[
                "Review exploration questions in exploration/questions.md",
                "Start exploration session with 'start_exploration' tool",
                "Work through systematic exploration before moving to specification"
            ]
        ).model_dump()
        
    except Exception as error:
        logger.error(f"Failed to initialize project: {error}")
        return InitProjectOutputSchema(
            success=False,
            project_name=validated_args.name,
            project_path="",
            message=f"Failed to initialize project: {str(error)}",
            next_steps=["Check permissions and try again"]
        ).model_dump()


async def list_projects_handler(args: dict) -> dict:
    """List all Compass projects"""
    validated_args = ListProjectsInputSchema(**args)
    
    try:
        projects_dir = Path(validated_args.projects_dir)
        if not projects_dir.exists():
            return ListProjectsOutputSchema(
                projects=[],
                total=0
            ).model_dump()
        
        projects = []
        
        for item in projects_dir.iterdir():
            if item.is_dir():
                compass_file = item / ".compass.json"
                if compass_file.exists():
                    try:
                        metadata_data = json.loads(compass_file.read_text())
                        metadata = ProjectMetadata.model_validate(metadata_data)
                        
                        projects.append(ProjectSummary(
                            name=metadata.name,
                            path=str(item),
                            status=metadata.status,
                            current_phase=metadata.current_phase,
                            created=metadata.created
                        ))
                    except Exception:
                        # Skip invalid projects
                        continue
        
        # Sort by created date, newest first
        projects.sort(key=lambda p: p.created, reverse=True)
        
        return ListProjectsOutputSchema(
            projects=projects,
            total=len(projects)
        ).model_dump()
        
    except Exception as error:
        logger.error(f"Failed to list projects: {error}")
        return ListProjectsOutputSchema(
            projects=[],
            total=0
        ).model_dump()


async def get_project_status_handler(args: dict) -> dict:
    """Get detailed project status"""
    validated_args = GetProjectStatusInputSchema(**args)
    
    try:
        projects_dir = Path(validated_args.projects_dir)
        project_path = projects_dir / validated_args.project_name
        
        if not project_path.exists():
            return GetProjectStatusOutputSchema(
                project_name=validated_args.project_name,
                current_phase="unknown",
                status_summary="Project not found",
                exploration_tasks=[],
                next_action="Initialize project first",
                recommendations=["Use 'init_project' tool to create this project"]
            ).model_dump()
        
        project_manager = ProjectManager(str(project_path))
        metadata = await project_manager.get_project_metadata()
        
        # Get exploration tasks
        exploration_tasks = []
        try:
            tasks_data = await project_manager.parse_tasks_file()
            exploration_tasks = [
                TaskSummary(text=task.text, completed=task.completed)
                for task in tasks_data.exploration_tasks
            ]
        except Exception:
            pass
        
        # Generate status summary
        completed_tasks = len([t for t in exploration_tasks if t.completed])
        total_tasks = len(exploration_tasks)
        
        if metadata.current_phase == "exploration":
            if total_tasks == 0:
                status_summary = "Exploration not started - no tasks defined yet"
                next_action = "Start exploration session to begin systematic discovery"
                recommendations = [
                    "Use 'start_exploration' tool to begin guided exploration",
                    "Review questions.md for exploration guidance",
                    "Focus on understanding the problem deeply before solutions"
                ]
            elif completed_tasks < total_tasks * 0.8:
                status_summary = f"Exploration in progress ({completed_tasks}/{total_tasks} tasks completed)"
                next_action = "Continue exploration sessions to complete remaining tasks"
                recommendations = [
                    "Focus on completing unchecked exploration tasks",
                    "Ensure each topic is thoroughly explored",
                    "Document insights and decisions in exploration sessions"
                ]
            else:
                status_summary = f"Exploration nearly complete ({completed_tasks}/{total_tasks} tasks completed)"
                next_action = "Complete remaining exploration or move to specification"
                recommendations = [
                    "Complete final exploration tasks",
                    "Review all exploration content for completeness",
                    "Use 'complete_exploration_phase' when ready to proceed"
                ]
        else:
            status_summary = f"Project in {metadata.current_phase} phase"
            next_action = f"Continue {metadata.current_phase} activities"
            recommendations = [f"Focus on {metadata.current_phase} phase requirements"]
        
        return GetProjectStatusOutputSchema(
            project_name=validated_args.project_name,
            current_phase=metadata.current_phase,
            status_summary=status_summary,
            exploration_tasks=exploration_tasks,
            next_action=next_action,
            recommendations=recommendations
        ).model_dump()
        
    except Exception as error:
        logger.error(f"Failed to get project status: {error}")
        return GetProjectStatusOutputSchema(
            project_name=validated_args.project_name,
            current_phase="error",
            status_summary=f"Error retrieving status: {str(error)}",
            exploration_tasks=[],
            next_action="Check project and try again",
            recommendations=["Ensure project exists and is properly initialized"]
        ).model_dump()


# Handler mapping
project_handlers = {
    "init_project": init_project_handler,
    "list_projects": list_projects_handler,
    "get_project_status": get_project_status_handler,
}